helix-core/src/line_ending.rs

```diff
@@ -0,0 +1,74 @@
+use crate::{Rope, RopeGraphemes, RopeSlice};
+
+/// Represents one of the valid Unicode line endings.
+#[derive(PartialEq, Copy, Clone, Debug)]
+pub enum LineEnding {
+    Crlf, // CarriageReturn followed by LineFeed
+    LF,   // U+000A -- LineFeed
+    VT,   // U+000B -- VerticalTab
+    FF,   // U+000C -- FormFeed
+    CR,   // U+000D -- CarriageReturn
+    Nel,  // U+0085 -- NextLine
+    LS,   // U+2028 -- Line Separator
+    PS,   // U+2029 -- ParagraphSeparator
+}
+
+pub fn rope_slice_to_line_ending(g: &RopeSlice) -> Option<LineEnding> {
+    if let Some(text) = g.as_str() {
+        str_to_line_ending(text)
+    } else if g == "\u{000D}\u{000A}" {
+        Some(LineEnding::Crlf)
+    } else {
+        // Not a line ending
+        None
+    }
+}
+
+pub fn str_to_line_ending(g: &str) -> Option<LineEnding> {
+    match g {
+        "\u{000D}\u{000A}" => Some(LineEnding::Crlf),
+        "\u{000A}" => Some(LineEnding::LF),
+        "\u{000B}" => Some(LineEnding::VT),
+        "\u{000C}" => Some(LineEnding::FF),
+        "\u{000D}" => Some(LineEnding::CR),
+        "\u{0085}" => Some(LineEnding::Nel),
+        "\u{2028}" => Some(LineEnding::LS),
+        "\u{2029}" => Some(LineEnding::PS),
+
+        // Not a line ending
+        _ => None,
+    }
+}
+
+pub fn auto_detect_line_ending(doc: &Rope) -> Option<LineEnding> {
+    // based on https://github.com/cessen/led/blob/27572c8838a1c664ee378a19358604063881cc1d/src/editor/mod.rs#L88-L162
+
+    let mut ending = None;
+    for line in doc.lines().take(1) {
```

Comment:
I remembered now why I had it checking the first 100 lines rather than just the first line.  Some line endings (e.g. FormFeed and ParagraphSeparator) are actually intended for mixed use with things like LF.

In practice it probably doesn't matter, since pretty much no one uses those anymore.  But it doesn't hurt to check, say, the first 10 lines or something just in case.

In fact, it might be worth excluding these three from the allowed auto-detect set:
* VT
* FF
* PS

Because I believe they're all special-use.  They should still be treated as line breaks, for correct text rendering, so they still need to be handled elsewhere in the code,  But they probably(?) never makes sense as a default line break.  I can investigate that more (I don't think it should block the PR).

Reply:
If we are indeed ignoring VT, FF, and PS, then it probably makes sense to keep iterating through the lines until we find a line ending that's not one of those.  And if we don't find one after some maximum number of lines (just to put a bounds on the extremely unlikely worst-case), then return None.